### PR TITLE
added - main.js for Node JS Sample for Batch Transcription #958

### PR DESCRIPTION
**Pull request by namrata01Apr**
## Purpose
This code was created for a customer request for sample for node js batch transcription . API for Model is still pending.

## Pull Request Type
What kind of change does this Pull Request introduce?
Create batch transcription sample for node.js

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: samples/batch/js/node/main.js [hunk 1]
```diff
@@ -0,0 +1,333 @@
+
+
+const { TranscriptionProperties } = require('speech_to_text_api_v30');
+var SpeechToTextApiV30 = require('speech_to_text_api_v30');
+var sleep = require('system-sleep');
+var defaultClient = SpeechToTextApiV30.ApiClient.instance;
+const request = require('request');
+
+   //Create API Instance
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:9]: nit: no indent

File: samples/batch/js/node/main.js [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+const NAME = "Simple transcription"
+const DESCRIPTION = "Simple transcription description"
+
+const LOCALE = "en-US"
+const RECORDINGS_BLOB_URI = ""
+
+// Provide the uri of a container with audio files for transcribing all of them with a single request
+const RECORDINGS_CONTAINER_URI = ""
+
+//Set model information when doing transcription with custom models
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:25]: nit: please always put a single space after `//`

File: samples/batch/js/node/main.js [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    
+    console.log (" url " + url);
+     if(file_data.kind!='Transcription')
+        continue;
+     else{
+        request(url, { json: true }, (err, res, body) => {
+            if (err) { 
+                return console.log(err); 
+            }
+         //   console.log("***************************************************************  response "+JSON.stringify(body));
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:100]: Please remove all commented code (except for features that are optional in the sample)

File: package-lock.json [hunk 4]
```diff
@@ -0,0 +1,178 @@
+{
```
Comment (chlandsi) [inline package-lock.json:1]: is this really needed? If yes, please move alongside with the code.

File: samples/batch/js/node/main.js [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+const LOCALE = "en-US"
+const RECORDINGS_BLOB_URI = ""
+
+// Provide the uri of a container with audio files for transcribing all of them with a single request
+const RECORDINGS_CONTAINER_URI = ""
+
+//Set model information when doing transcription with custom models
+const MODEL_REFERENCE = 'None' ; // guid of a custom model
+
+var callback2 = function(error, data, response) {
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:28]: this is unused?

File: samples/batch/js/node/main.js [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+          {
+           
+          }
+        else if (trans_status == "Failed")
+            console.log("Transcription failed: {transcription.properties.error.message}")
+
+        }
+      };
+
+ var callback = function(error, data, response) {
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:52]: this is unused?

File: samples/batch/js/node/main.js [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+          },
+          "locale": LOCALE,
+          "displayName": NAME + new Date(),
+          "description" : DESCRIPTION
+        };
+        return opts;
+}//end of _transcribeFromSingleBlob
+
+
+//To be worked on
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:248]: remove?

File: samples/batch/js/node/main.js [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    apiKeyHeader.apiKey = '';
+    // Configure API key authorization: apiKeyQuery
+    var apiKeyQuery = defaultClient.authentications['apiKeyQuery'];
+    apiKeyQuery.apiKey = '';
+
+    //Create API Instance
+    //var apiInstance = new SpeechToTextApiV30.DefaultApi();
+
+    const subscriptionKey = "{your-subscription-key-here}";
+    const serviceRegion = "centralindia"; // e.g., "westus"
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:129]: use the constants defined at the top. Also elsewhere.

File: samples/batch/js/node/main.js [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    defaultClient.basePath = 'https://centralindia.api.cognitive.microsoft.com/speechtotext/v3.0'.replace(/\/+$/, '');
+
+    // get all transcriptions for the subscription
+     var opts = { 
+        'skip': 56, // Number | Format - int32. Number of transcriptions that will be skipped.
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:177]: why these numbers?

File: samples/batch/js/node/package.json [hunk 10]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "visual-analyzer",
+  "version": "0.1.0",
+  "private": true,
+  "homepage": "/build",
+  "dependencies": {
+    "@azure/cognitiveservices-computervision": "^7.1.0",
```
Comment (chlandsi) [inline samples/batch/js/node/package.json:7]: is this file necessary? some of the dependencies are unrelated. Also please avoid local paths.

File: samples/batch/js/node/package-lock.json [hunk 11]
```diff
@@ -0,0 +1,554 @@
+{
+  "name": "visual-analyzer",
```
Comment (chlandsi) [inline samples/batch/js/node/package-lock.json:2]: is this file necessary?

**Review by chlandsi — CHANGES_REQUESTED** [review]
Thank you for the sample; however, this still seems incomplete and/or needs some cleanup in order to be a general sample. Please have a look at the comments for some of the issues.

File: samples/batch/js/node/main.js (hunk 1 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: samples/batch/js/node/main.js (hunk 2 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: samples/batch/js/node/main.js (hunk 3 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: package-lock.json (hunk 4 above)
Comment (namrata01Apr) [reply package-lock.json:1]: Not needed

File: samples/batch/js/node/main.js (hunk 7 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: samples/batch/js/node/main.js (hunk 8 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: samples/batch/js/node/main.js (hunk 9 above)
Comment (namrata01Apr) [reply samples/batch/js/node/main.js]: @chlandsi - Have incorporated the changes in the latest code for main.js.

File: samples/batch/js/node/package-lock.json (hunk 11 above)
Comment (namrata01Apr) [reply samples/batch/js/node/package-lock.json:2]: Not necessary

**Comment by namrata01Apr** [comment]
*@chlandsi*  Request you to pls review and help with publishing the code.

File: samples/batch/js/node/main.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+  } else {
+   console.log('callbackGetTransFiles API called successfully. Returned data: ' + JSON.stringify(data));
+   pag_files = data.values;
+   for (var i = 0; i < pag_files.length; i++)
+   {
+     var file_data = pag_files[i];
+     var url = file_data.links.contentUrl; 
+     if(file_data.kind !='Transcription')
+        continue;
+     else{
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:54]: nit: please use consistent indentation and spacing before/after brackets and operators.

File: samples/batch/js/node/main.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+            }
+        });
+        
+     }
+   }
+  }
+};
+
+// uncomment to execute 
+// _transcribe();
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:71]: uncomment

File: samples/batch/js/node/main.js [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+{
+    
+    // Configure API key authorization: apiKeyHeader
+    var apiKeyHeader = defaultClient.authentications['apiKeyHeader'];
+    apiKeyHeader.apiKey = API_KEY;
+    // Configure API key authorization: apiKeyQuery
+    var apiKeyQuery = defaultClient.authentications['apiKeyQuery'];
+    apiKeyQuery.apiKey = API_KEY;
+
+    defaultClient.basePath = 'https:// centralindia.api.cognitive.microsoft.com/speechtotext/v3.0'.replace(/\/+$/, '');
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:84]: space in Uri?

Comment (chlandsi) [inline samples/batch/js/node/main.js:84]: Region needs to be defined as a constant at the top of the file. Make sure no region is hardcoded anywhere in the code.

Comment (chlandsi) [inline samples/batch/js/node/main.js:84]: why the replace(.)?

File: samples/batch/js/node/main.js [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+    // Uncomment this block to transcribe one file.
+    // var opts = _transcribeFromSingleBlob(RECORDINGS_BLOB_URI);
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:87]: please make this the default for the sample

Comment (chlandsi) [reply samples/batch/js/node/main.js:87]: and adapt the comment. Also mention that only one method of transcription can be used.

File: samples/batch/js/node/main.js [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+  // Delete all transcriptions associated with your speech resource.
+  console.log("Deleting all existing completed transcriptions.")
+  // Configure API key authorization: apiKeyHeader
+  var apiKeyHeader = defaultClient.authentications['apiKeyHeader'];
+  apiKeyHeader.apiKey = API_KEY;
+  // Configure API key authorization: apiKeyQuery
+  var apiKeyQuery = defaultClient.authentications['apiKeyQuery'];
+  apiKeyQuery.apiKey = API_KEY;
+  
+  defaultClient.basePath = 'https:// centralindia.api.cognitive.microsoft.com/speechtotext/v3.0'.replace(/\/+$/, '');
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:119]: hardcoded region

File: samples/batch/js/node/main.js [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+  var opts = { 
+  };
+  var callback = function(error, data, response) {
+        if (error) {
+          console.error(error);
+        } else {
+          // uncomment to display the response of Delete All transcriptions
+          // console.log('Delete all transcriptions API called successfully. Returned data: ' + JSON.stringify(data));
+          pag_files = data.values;
+          console.log(pag_files.length);
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:131]: please add a message

File: samples/batch/js/node/main.js [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+          console.log("Transcription ID created is : "+tid);
+          apiInstance.getTranscription(tid,callbackGetTrans);
+          
+        }
+      });
+    
+}// end of _transcribe()
+
+
+function _deleteAllTranscriptions(){
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:108]: This does not actually delete all transcriptions, because the GET transcriptions API returns transcription as pages (of 100 entities by default). Please mention this in a comment, and that it might be necessary to call this function multiple times; or implement the paged call.

File: samples/batch/js/node/main.js [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+          "contentContainerUrl": uri,
+          "properties": {
+          "wordLevelTimestampsEnabled": true
+          },
+          "locale": LOCALE,
+          "displayName": NAME + new Date(),
+          "description" : DESCRIPTION
+        };
+        return opts;
+}// end of _transcribeFromSingleBlob
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:204]: update

File: package-lock.json (hunk 4 above)
Comment (chlandsi) [reply package-lock.json:1]: Please remove the file then.

File: samples/batch/js/node/package-lock.json (hunk 11 above)
Comment (chlandsi) [reply samples/batch/js/node/package-lock.json:2]: Please remove the file then.

**Review by chlandsi — CHANGES_REQUESTED** [review]

File: samples/batch/js/node/main.js [hunk 20]
```diff
@@ -0,0 +1,236 @@
+const { TranscriptionProperties } = require('speech_to_text_api_v30');
```
Comment (chlandsi) [inline samples/batch/js/node/main.js:1]: Please also add a license header, similar to the other samples.

**Comment by namrata01Apr** [comment]
*@chlandsi* requested changes have now been made. Do let me know if you
find anything further

File: package-lock.json (hunk 4 above)
Comment (chlandsi) [reply package-lock.json:1]: This still needs to be removed.

File: samples/batch/README.md [hunk 21]
```diff
@@ -11,3 +11,4 @@ Available samples:
 | C# | [csharp](csharp) | C# calling batch transcription REST API through System.Net.Http |
 | C# | [Transcription Enabled Storage](transcription-enabled-storage) | Project to automatically transcribe all audio files which are added to an Azure Storage Container. Setup via ARM template. |
 | Python | [python](python) | Python client calling batch transcription REST API |
+| Node.js | [nodejs](nodejs) | Node.js client calling batch transcription REST API |
```
Comment (chlandsi) [inline samples/batch/README.md:14]: replace (nodejs) by (js/node) to avoid a broken link

File: samples/batch/js/node/README.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+```bash
+npm install system-sleep
+npm install request
+```
+
+## Locally modify the Swagger client for the issue related to 'File'
+Modified node_modules/speech_to_text_api_v30/src/model/PaginatedFiles.constructFromObject as it was throwing an error reference File not found.
+
+	Was : obj.values = ApiClient.convertToType(data['values'], [File]);
```
Comment (chlandsi) [inline samples/batch/js/node/README.md:31]: indentation

**Review by chlandsi — CHANGES_REQUESTED** [review]

**Comment by namrata01Apr** [comment]
*@chlandsi* changes have been submitted,  though i dont see the package
json file. Do let me know if you are still seeing this.

**Comment by chlandsi** [comment]
Hi @namrata01Apr, yes, the file is still part of the PR (look in the "Files changed" tab of the PR). Note the file is in the root of the repo, not the samples directory.

**Comment by namrata01Apr** [comment]
@chlandsi ,
Ok. i have that deleted now.

**Review by chlandsi — APPROVED** [review]
Looks good now, thanks @namrata01Apr !

**Comment by namrata01Apr** [comment]
Thank *@chlandsi .  *
this is new to me, if you could confirm what is the next step to publish
the same ? Is there any action needed from me?

**Comment by chlandsi** [comment]
Hi @namrata01Apr, I merged the PR. Thank you for the contribution.
